// SPDX-FileCopyrightText: 2025 Carnegie Mellon University - LivingEdgeLab
//
// SPDX-License-Identifier: GPL-2.0-only

syntax = "proto3";
package steeleagle.protocol.services.control_service;

import "common.proto";
import "google/protobuf/duration.proto";

/*
 * Used for low-level control of a vehicle.
 *
 * This service is hosted by the driver module and represents the global
 * control interface for the vehicle. Most methods called here will result
 * in actuation of the vehicle if it is armed (be careful!). Some methods,
 * like TakeOff, may take some time to complete. For this reason, it is
 * not advisable to set a timeout/deadline on the RPC call. However, to
 * ensure that the service is progressing, a client can either check
 * telemetry or listen for `IN_PROGRESS` response heartbeats which are
 * streamed back from the RPC while executing an operation.
 */
service Control {
  /*
   * Connect to the vehicle.
   *
   * Connects to the underlying vehicle hardware. Generally, this 
   * method is called by the law authority on startup and is not
   * called by user code.
   */
  rpc Connect (ConnectRequest)
	returns (steeleagle.protocol.common.Response) {}
  /*
   * Disconnect from the vehicle.
   * 
   * Disconnects from the underlying vehicle hardware. Generally,
   * this method is called by the law authority when it attempts
   * a driver restart and is not called by user code.
   */
  rpc Disconnect (DisconnectRequest) 
	returns (steeleagle.protocol.common.Response) {}
  /*
   * Order the vehicle to arm.
   *
   * Arms the vehicle. This is required before any other commands
   * are run, otherwise the methods will return `FAILED_PRECONDITION`.
   * Once the vehicle is armed, all subsequent actuation methods
   * _will move the vehicle_. Make sure to go over the manufacturer
   * recommended vehicle-specific pre-operation checklist before arming.
   */
  rpc Arm (ArmRequest) 
	returns (steeleagle.protocol.common.Response) {}
  /*
   * Order the vehicle to disarm.
   *
   * Disarms the vehicle. Prevents any further actuation methods
   * from executing, unless the vehicle is re-armed.
   */
  rpc Disarm (DisarmRequest) 
	returns (steeleagle.protocol.common.Response) {}
  /*
   * Send a joystick command to the vehicle.
   *
   * Causes the vehicle to accelerate towards a provided velocity
   * setpoint over a provided duration. This is useful for fine-grained 
   * control based on streamed datasink results or for tele-operating 
   * the vehicle from a remote commander.
   */
  rpc Joystick (JoystickRequest) 
	returns (steeleagle.protocol.common.Response) {}
  /*
   * Order the vehicle to take off.
   *
   * Causes the vehicle to take off to a specified take off altitude.
   * If the vehicle is not a UAV, this method will be unimplemented.
   */
  rpc TakeOff (TakeOffRequest) 
	returns (stream steeleagle.protocol.common.Response) {}
  /*
   * Order the vehicle to land.
   *
   * Causes the vehicle to land at its current location. If the 
   * vehicle is not a UAV, this method will be unimplemented.
   */
  rpc Land (LandRequest) 
	returns (stream steeleagle.protocol.common.Response) {}
  /*
   * Order the vehicle to hold/loiter.
   *
   * Causes the vehicle to hold at its current location and to
   * cancel any ongoing movement commands (`ReturnToHome` e.g.).
   */
  rpc Hold (HoldRequest) 
	returns (stream steeleagle.protocol.common.Response) {}
  /*
   * Orders an emergency shutdown of the vehicle motors.
   *
   * Causes the vehicle to immediately turn off its motors. _If the 
   * vehicle is a UAV, this will result in a freefall_. Use this
   * method only in emergency situations.
   */
  rpc Kill (KillRequest) 
	returns (stream steeleagle.protocol.common.Response) {}
  /*
   * Set the home location of the vehicle.
   *
   * Changes the home location of the vehicle. Future `ReturnToHome`
   * commands will move the vehicle to the provided location instead
   * of its starting position.
   */
  rpc SetHome (SetHomeRequest) 
	returns (steeleagle.protocol.common.Response) {}
  /*
   * Order the vehicle to return to its home position.
   *
   * Causes the vehicle to return to its home position. If the home position 
   * has not been explicitly set, this will be its start position (defined 
   * as its takeoff position for UAVs). If the home position has been 
   * explicitly set, by `SetHome`, the vehicle will return to that 
   * position instead.
   */
  rpc ReturnToHome (ReturnToHomeRequest) 
	returns (stream steeleagle.protocol.common.Response) {}
  /*
   * Order the vehicle to move to a global position.
   *
   * Causes the vehicle to transit to the provided global position. The vehicle
   * will interpret the heading of travel according to `heading_mode`:
   * - `TO_TARGET` -> turn to face the target position bearing
   * - `HEADING_START` -> turn to face the provided heading in the global position object. 
   * 
   * This will be the heading the vehicle maintains for the duration of transit. 
   * Generally only UAVs will support `HEADING_START`.
   * 
   * The vehicle will move towards the target at the specified maximum velocity 
   * until the vehicle has reached its destination. Error tolerance is determined 
   * by the driver. Maximum velocity is interpreted from `max_velocity` as follows:
   * - `x_vel` -> maximum _horizontal_ velocity
   * - `y_vel` -> ignored
   * - `z_vel` -> maximum _vertical_ velocity _(UAV only)_
   *
   * If no maximum velocity is provided, the driver will use a preset speed usually 
   * determined by the manufacturer or hardware settings.
   *
   * _(UAV only)_ During motion, the vehicle will also ascend or descend towards the 
   * target altitude, linearly interpolating this movement over the duration of
   * travel. The vehicle will interpret altitude from `altitude_mode` as follows:
   * - `ABSOLUTE` -> altitude is relative to MSL (Mean Sea Level)
   * - `RELATIVE` -> altitude is relative to take off position
   */
  rpc SetGlobalPosition (SetGlobalPositionRequest)
	returns (stream steeleagle.protocol.common.Response) {}
  /*
   * Order the vehicle to move to a relative position.
   *
   * Causes the vehicle to transit to the provided relative position. The vehicle
   * will interpret the input position according to `frame` as follows:
   * - `BODY` -> (`x`, `y`, `z`) = (forward offset, right offset, up offset) _from current position_
   * - `NEU` -> (`x`, `y`, `z`) = (north offset, east offset, up offset) _from start position_
   * 
   * The vehicle will move towards the target at the specified maximum velocity 
   * until the vehicle has reached its destination. Error tolerance is determined 
   * by the driver. Maximum velocity is interpreted from `max_velocity` as follows:
   * - `x_vel` -> maximum _horizontal_ velocity
   * - `y_vel` -> ignored
   * - `z_vel` -> maximum _vertical_ velocity _(UAV only)_
   * 
   * If no maximum velocity is provided, the driver will use a preset speed usually 
   * determined by the manufacturer or hardware settings.
   */
  rpc SetRelativePosition (SetRelativePositionRequest)
	returns (stream steeleagle.protocol.common.Response) {}
  /*
   * Order the vehicle to accelerate to a velocity.
   *
   * Causes the vehicle to accelerate until it reaches a provided velocity.
   * Error tolerance is determined by the driver. The vehicle will interpret 
   * the input velocity according to `frame` as follows:
   * - `BODY` -> (`x_vel`, `y_vel`, `z_vel`) = (forward velocity, right velocity, up velocity)
   * - `NEU` -> (`x_vel`, `y_vel`, `z_vel`) = (north velocity, east velocity, up velocity)
   */
  rpc SetVelocity (SetVelocityRequest)
	returns (stream steeleagle.protocol.common.Response) {}
  /*
   * Order the vehicle to set a new heading.
   *
   * Causes the vehicle to turn to face the provided global position. The vehicle
   * will interpret the final heading according to `heading_mode`:
   * - `TO_TARGET` -> turn to face the target position bearing
   * - `HEADING_START` -> turn to face the provided heading in the global position object. 
   */
  rpc SetHeading (SetHeadingRequest)
	returns (stream steeleagle.protocol.common.Response) {}
  /*
   * Order the vehicle to set the pose of a gimbal.
   *
   * Causes the vehicle to actuate a gimbal to a new pose. The vehicle
   * will interpret the new pose type from `pose_mode` as follows: 
   * - `ABSOLUTE` -> absolute angle
   * - `RELATIVE` -> angle relative to current position
   * - `VELOCITY` -> angular velocities
   * 
   * The vehicle will interpret the new pose angles according to `frame` 
   * as follows:
   * - `BODY` -> (`pitch`, `roll`, `yaw`) = (body pitch, body roll, body yaw)
   * - `NEU` -> (`pitch`, `roll`, `yaw`) = (body pitch, body roll, global yaw)
   */
  rpc SetGimbalPose (SetGimbalPoseRequest)
	returns (stream steeleagle.protocol.common.Response) {}
  /*
   * Configure the vehicle imaging stream.
   *
   * Sets which imaging sensors are streaming and sets their target
   * frame rates.
   */
  rpc ConfigureImagingSensorStream (ConfigureImagingSensorStreamRequest)
	returns (steeleagle.protocol.common.Response) {}
  /*
   * Configure the vehicle telemetry stream.
   *
   * Sets the frequency of the telemetry stream.
   */
  rpc ConfigureTelemetryStream (ConfigureTelemetryStreamRequest)
	returns (steeleagle.protocol.common.Response) {}   
}

message ConnectRequest {
  steeleagle.protocol.common.Request request = 1; // request data
}

message DisconnectRequest {
  steeleagle.protocol.common.Request request = 1; // request data
}

message ArmRequest {
  steeleagle.protocol.common.Request request = 1; // request data
}

message DisarmRequest {
  steeleagle.protocol.common.Request request = 1; // request data
}

message JoystickRequest {
  steeleagle.protocol.common.Request request = 1; // request data
  // target velocity to move towards
  steeleagle.protocol.common.Velocity velocity = 2;
  // time of actuation after which the vehicle will Hold
  google.protobuf.Duration duration = 3;
}

message TakeOffRequest {
  steeleagle.protocol.common.Request request = 1; // request data
  // take off height in relative altitude [meters]
  float take_off_altitude = 2;
}

message LandRequest {
  steeleagle.protocol.common.Request request = 1; // request data
}

message HoldRequest {
  steeleagle.protocol.common.Request request = 1; // request data
}

message KillRequest {
  steeleagle.protocol.common.Request request = 1; // request data
}

message SetHomeRequest {
  steeleagle.protocol.common.Request request = 1; // request data
  steeleagle.protocol.common.Location location = 2; // new home location
}

message ReturnToHomeRequest {
  steeleagle.protocol.common.Request request = 1; // request data
}

// Altitude mode switch.
enum AltitudeMode {
  ABSOLUTE = 0; // meters above Mean Sea Level
  RELATIVE = 1; // meters above takeoff position
}

// Heading mode switch.
enum HeadingMode {
  TO_TARGET = 0; // orient towards the target location
  HEADING_START = 1; // orient towards the given heading
}

message SetGlobalPositionRequest {
  steeleagle.protocol.common.Request request = 1; // request data
  steeleagle.protocol.common.Location location = 2; // target global position
  // determines how the vehicle will orient during transit (default: `TO_TARGET`)
  optional HeadingMode heading_mode = 4;
  // determines how the vehicle will interpret altitude (default: `ABSOLUTE`)
  optional AltitudeMode altitude_mode = 5;
  // maximum velocity during transit
  optional steeleagle.protocol.common.Velocity max_velocity = 6;
}

// Reference frame mode switch.
enum ReferenceFrame {
  BODY = 0; // vehicle reference frame
  NEU = 1; // NEU (North, East, Up) reference frame
}

message SetRelativePositionRequest {
  steeleagle.protocol.common.Request request = 1; // request data
  // target relative position
  steeleagle.protocol.common.Position position = 2;
  // maximum velocity during transit
  optional steeleagle.protocol.common.Velocity max_velocity = 3;
  // frame of reference
  optional ReferenceFrame frame = 4;
}

message SetVelocityRequest {
  steeleagle.protocol.common.Request request = 1; // request data
  // target velocity
  steeleagle.protocol.common.Velocity velocity = 2;
  // frame of reference
  optional ReferenceFrame frame = 3;
}

message SetHeadingRequest {
  steeleagle.protocol.common.Request request = 1; // request data
  // target heading or global location to look at
  steeleagle.protocol.common.Location location = 2;
  // determines how the drone will orient
  optional HeadingMode heading_mode = 5;
}

// Pose mode switch.
enum PoseMode {
  ANGLE = 0; // absolute angle
  OFFSET = 1; // request data // Offset from current
  VELOCITY = 2; // rotational velocities
}

message SetGimbalPoseRequest {
  steeleagle.protocol.common.Request request = 1; // request data
  uint32 gimbal_id = 2; // ID of the target gimbal
  steeleagle.protocol.common.Pose pose = 3; // target pose
  optional PoseMode pose_mode = 4; // specifies how to interpret the target pose
  // frame of reference
  optional ReferenceFrame frame = 5;
}

// Configuration for an imaging sensor.
message ImagingSensorConfiguration {
  uint32 id = 1; // target imaging sensor ID
  bool set_primary = 2; // set this sensor as the primary stream
  uint32 set_fps = 3; // target FPS for stream
} 

message ConfigureImagingSensorStreamRequest {
  steeleagle.protocol.common.Request request = 1; // request data
  // list of configurations to be updated
  repeated ImagingSensorConfiguration configurations = 2;
}

message ConfigureTelemetryStreamRequest {
  steeleagle.protocol.common.Request request = 1; // request data
  // target frequency of telemetry generation, in Hz
  uint32 frequency = 2;
}
